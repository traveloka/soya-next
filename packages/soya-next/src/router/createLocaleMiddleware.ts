import getBaseUrl from "../utils/getBaseUrl";

import type { RequestHandler } from "express";

export interface CreateLocaleMiddlewareOption {
  defaultLocale?: string;
  siteLocales?: string[];
}

export default function createLocaleMiddleware(
  options?: CreateLocaleMiddlewareOption
): RequestHandler {
  const { defaultLocale, siteLocales } = options || {};

  if (typeof defaultLocale !== "string") {
    throw new Error("Expected defaultLocale to be a locale string.");
  }
  if (
    typeof siteLocales !== "object" ||
    siteLocales.constructor !== Array ||
    siteLocales.length === 0
  ) {
    throw new Error("Expected siteLocales to be an array of locale string.");
  }
  return (req, res, next) => {
    let [language, country] = defaultLocale.split("-");
    const baseUrl = getBaseUrl(req);
    const { pathname } = new URL(req.url, baseUrl.origin);
    const [localeSegment] = pathname.substr(1).split("/");
    if (localeSegment) {
      const [languageSegment, countrySegment = country] = localeSegment.split(
        "-"
      );
      if (siteLocales.indexOf(`${languageSegment}-${countrySegment}`) !== -1) {
        country = countrySegment;
        language = languageSegment;
        req.url = req.url.substr(localeSegment.length + 1) || "/";
      }
    }
    req.defaultLocale = defaultLocale;
    req.siteLocales = siteLocales;
    req.locale = {
      country,
      language,
    };
    next();
  };
}
