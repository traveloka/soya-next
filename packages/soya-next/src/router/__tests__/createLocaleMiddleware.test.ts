import createLocaleMiddleware from "../createLocaleMiddleware";

describe("Create Locale Middleware", () => {
  function setupTest() {
    return {
      options: {
        defaultLocale: "id-id",
        siteLocales: ["id-id", "en-id", "en-sg"],
      },
      res: {} as any,
      next: jest.fn(),
    };
  }

  it("requires default locale", () => {
    expect(() => createLocaleMiddleware()).toThrowErrorMatchingSnapshot();
  });

  it("requires site locales", () => {
    expect(() =>
      createLocaleMiddleware({ defaultLocale: "id-id" })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should fallback to default locale", () => {
    const { options, next, res } = setupTest();
    const req: any = { url: "/" };
    createLocaleMiddleware(options)(req, res, next);
    expect(req).toMatchSnapshot();
  });

  it("should fallback country to its default and remove locale segment from url", () => {
    const { options, next, res } = setupTest();
    const req: any = { url: "/en/" };
    createLocaleMiddleware(options)(req, res, next);
    expect(req).toMatchSnapshot();
  });

  it("should match even without trailing slash", () => {
    const { options, next, res } = setupTest();
    const req: any = { url: "/en-sg" };
    createLocaleMiddleware(options)(req, res, next);
    expect(req).toMatchSnapshot();
  });

  it("should match available site locales and remove locale segment from url", () => {
    const { options, next, res } = setupTest();
    const req: any = { url: "/en-sg/" };
    createLocaleMiddleware(options)(req, res, next);
    expect(req).toMatchSnapshot();
  });

  it("should fallback to default locale if none match", () => {
    const { options, next, res } = setupTest();
    const req: any = { url: "/ms-my/" };
    createLocaleMiddleware(options)(req, res, next);
    expect(req).toMatchSnapshot();
  });
});
