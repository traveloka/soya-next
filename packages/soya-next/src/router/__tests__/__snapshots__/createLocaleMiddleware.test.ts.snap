// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Locale Middleware requires default locale 1`] = `"Expected defaultLocale to be a locale string."`;

exports[`Create Locale Middleware requires site locales 1`] = `"Expected siteLocales to be an array of locale string."`;

exports[`Create Locale Middleware should fallback country to its default and remove locale segment from url 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "en",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
    "en-sg",
  ],
  "url": "/",
}
`;

exports[`Create Locale Middleware should fallback to default locale 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "id",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
    "en-sg",
  ],
  "url": "/",
}
`;

exports[`Create Locale Middleware should fallback to default locale if none match 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "id",
    "language": "id",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
    "en-sg",
  ],
  "url": "/ms-my/",
}
`;

exports[`Create Locale Middleware should match available site locales and remove locale segment from url 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "sg",
    "language": "en",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
    "en-sg",
  ],
  "url": "/",
}
`;

exports[`Create Locale Middleware should match even without trailing slash 1`] = `
Object {
  "defaultLocale": "id-id",
  "locale": Object {
    "country": "sg",
    "language": "en",
  },
  "siteLocales": Array [
    "id-id",
    "en-id",
    "en-sg",
  ],
  "url": "/",
}
`;
